File: 662-maximum-width-of-binary-tree/leochang/662.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        int ans = 0;
        unordered_map<int, int> leftest;
        dfs(root, ans, 1, 1, leftest);
        return ans;
    }
    void dfs(TreeNode *node, int &ans, int id, int level, unordered_map<int, int> &leftest) {
        if (node == NULL) {
            return;
        }
        if (leftest.size() < level) {
            leftest[level] = id;
        }

        ans = max(ans, id - leftest[level]+1);
        dfs(node->left, ans, id*2, level+1, leftest);
        dfs(node->right, ans, id*2+1, level+1, leftest);
    }
};

File: 662-maximum-width-of-binary-tree/ken/662.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def widthOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """

        cur = [(root, 0)]
        ans = 0
        while cur:
            next = []
            left = -1
            right = -1

            for i in xrange(len(cur)):
                node = cur[i][0]
                pos = cur[i][1]

                if node is None:
                    continue

                if left == -1:
                    left = pos

                right = pos
                next.append((node.left, pos * 2))
                next.append((node.right, pos * 2 + 1))

            ans = max(ans, right - left + 1)
            cur = next

        return ans
